docs/source/llms/notebooks/chat-model-tool-calling.ipynb

```diff
@@ -0,0 +1,467 @@
+    "2. `messages`: List\\[[ChatMessage](https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.llm.ChatMessage)\\]. This is the chat input that the model uses for generation.\n",
+    "3. `params`: [ChatParams](https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.llm.ChatParams). These are commonly used params used to configure the chat model, e.g. `temperature`, `max_tokens`, etc. This is where the tool specifications can be found.\n",
+    "\n",
+    "For the implementation, we'll simply forward the user's input to OpenAI, and provide the `get_weather` tool as an option for the LLM to use if it chooses to do so. If we receive a tool call request, we'll call the `get_weather()` function and return the response back to OpenAI. We'll need to use what we've defined in the previous two steps in order to do this."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "from typing import List\n",
+    "\n",
+    "from openai import OpenAI\n",
+    "\n",
+    "import mlflow\n",
+    "from mlflow.types.llm import (\n",
+    "    ChatMessage,\n",
+    "    ChatParams,\n",
+    "    ChatResponse,\n",
+    "    FunctionToolDefinition,\n",
+    "    ParamProperty,\n",
+    "    ToolParamsSchema,\n",
+    ")\n",
+    "\n",
+    "# replace this with your own MLflow tracking URI\n",
+    "mlflow.set_tracking_uri(\"http://localhost:5000\")\n",
+    "\n",
+    "\n",
+    "class WeatherModel(mlflow.pyfunc.ChatModel):\n",
+    "    def __init__(self):\n",
+    "        # from step 1 above, our tool definition\n",
+    "        weather_tool = FunctionToolDefinition(\n",
+    "            name=\"get_weather\",\n",
+    "            description=\"Get weather information\",\n",
+    "            parameters=ToolParamsSchema(\n",
+    "                {\n",
+    "                    \"city\": ParamProperty(\n",
+    "                        type=\"string\",\n",
+    "                        description=\"City name to get weather information for\",\n",
+    "                    ),\n",
+    "                }\n",
+    "            ),\n",
+    "        ).to_tool_definition()\n",
+    "\n",
+    "        # OpenAI expects tools to be provided as a list of dictionaries\n",
+    "        self.tools = [weather_tool.to_dict()]\n",
+    "\n",
+    "    # from step 2 above, the implementation of the tool\n",
+    "    def get_weather(self, city: str) -> str:\n",
+    "        return \"It's sunny in {}, with a temperature of 20C\".format(city)\n",
+    "\n",
+    "    # the core method that needs to be implemented. this function\n",
+    "    # will be called every time a user sends messages to our model\n",
+    "    def predict(self, context, messages: List[ChatMessage], params: ChatParams):\n",
+    "        # instantiate the OpenAI client\n",
+    "        client = OpenAI()\n",
```

Comment:
Let's move `client` in `__init__`. In this tutorial, it doesn't matter because `predict` is called a few times.

Reply:
it looks like the OpenAI client is not serializable, so it can't go in `init` unless i use the `load_context` stuff (which might take away some focus from the point of the tutorial)

Reply:
Got it. Let's this leave as it is.